pdudaemon/drivers/conrad197720.py

```diff
@@ -0,0 +1,354 @@
+        """
+        Init all cards
+        :param self: The object itself
+        :raises RuntimeError: all retries failed, communication error
+        """
+        self.__sendCommand(CMD_INIT, ADDR_FIRST_CARD, 0)
+        # make sure, all cards run with default setting
+        for card_addr in range (1,self.getNumCards()+1):
+            option = self.__sendCommand(CMD_GETOPTION, card_addr, 0)
+            if option!=OPTION_BROADCAST_SEND_AND_NO_BLOCK:
+                # set card to default setting
+                log.debug(f"set option to default for card: {card_addr}")
+                self.__sendCommand(CMD_SETOPTION, card_addr, OPTION_BROADCAST_SEND_AND_NO_BLOCK)
+
+    def updateSingle(self, command, port):
+        """
+        Update a single port of a card
+        :param self: The object itself
+        :param command: The command to execute, "on" "off" or "toggle"
+        :param port: The port number 0..n, all ports of all cards are in one range, e.g. port 9 is the second port of the second card
+        :return: True on success
+        :rtype: boolean
+        :raises RuntimeError: unknown command
+        :raises RuntimeError: no cards present
+        :raises RuntimeError: port number invalid
+        :raises RuntimeError: all retries failed, communication error
+        """
+        for i in range (0,RETRY_COUNT+1):
+            card_addr = int(port / PORTS_PER_CARD)+1
+            port_index = int(port % PORTS_PER_CARD)
+
+            if command=="off":
+                card_command=CMD_DELSINGLE
+            elif command=="on":
+                card_command=CMD_SETSINGLE
+            elif command=="toggle":
+                card_command=CMD_TOGGLESINGLE
+            else:
+                raise RuntimeError(f"Unknown command {command}")
+
+            if 0==self.getNumCards():
+                # retry init
+                self.init()
+                if 0==self.getNumCards():
+                    raise RuntimeError("No conrad197720 compatible cards present")
+
+            if port < 0 or card_addr > self.getNumCards():
+                raise RuntimeError(f"Port number {port} has to be between 0 and {(self.getNumCards() * PORTS_PER_CARD)-1}")
+
+            recv = self.__sendCommand(CMD_GETPORT, card_addr, 0)
+            old_state = (recv & 1 << port_index) >> port_index
+
+            recv = self.__sendCommand(card_command, card_addr, 1 << port_index)
+            new_state = (recv & 1 << port_index) >> port_index
+
+            log.debug (f"updateSingle card_addr {card_addr} port_index {port_index} command {command} state {old_state} -> {new_state}")
+
+            if (command=="off" and new_state==0) or \
+               (command=="on" and new_state==1) or \
+               (command=="toggle" and new_state!=old_state):
+                return True
+
+            log.debug ("updateSingle retry")
+
+        raise RuntimeError("All retries failed, communication error")
+
+    def updateCard(self, port_data, card_addr):
```

Comment:
This functions seems unused. Why does it exist? :)

Reply:
It is a function which could be used to reset all outputs of the card, e.g. after a reboot of the host. I did find it useful and i wanted to support it, if pdudaemon once supports this feature. But i will remove it for now, i can add it again later, if necessary.